### PR TITLE
Move to rust types generated from provenance protos

### LINKED ISSUE #101: Epic: Use protos instead of json for bindings
## Summary

The current provenance and cosmos bindings use json for serde. Moving to protos that can be sourced from provenance and cosmos would simplify the module side encoding/decoding process.

## Proposal

Investigate how osmosis has achieved this and implement similarly for provenance. This requires changes in `provwasm` as well as `provenance`.

____

#### For Admin Use

- [X] Not duplicate issue
- [X] Appropriate labels applied
- [X] Appropriate contributors tagged
- [X] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by kwtalley**
This PR deprecates the old style of bindings and replaces them with types generated using the provenance protos. It is a major change requiring contracts to migrate their code to use them. The old version will still be available for a release cycle to allow developers to continue without loss of functionality.

closes #101

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -32,11 +32,11 @@ jobs:
           cd ./contracts/tutorial
           make all
       - name: Smart Contract Test setup
-        uses: provenance-io/provenance-testing-action@34bfd03104a8a2ebd2e03f60c90353581cc5016a
+        uses: provenance-io/provenance-testing-action@eeeaf2792a9fbab3ccd74439a9e132d079fcdef5
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
-          provenance_version: "v1.14.1"
-          smart_contract_action_version: "34bfd03104a8a2ebd2e03f60c90353581cc5016a"
+          provenance_version: "v1.15.0-rc2"
```
Comment (Taztingo) [inline .github/workflows/test.yml:38]: Should this be v1.15.0?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -55,11 +55,11 @@ jobs:
           cd ./contracts/attrs
           make all
       - name: Smart Contract Test setup
-        uses: provenance-io/provenance-testing-action@34bfd03104a8a2ebd2e03f60c90353581cc5016a
+        uses: provenance-io/provenance-testing-action@eeeaf2792a9fbab3ccd74439a9e132d079fcdef5
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
-          provenance_version: "v1.14.1"
-          smart_contract_action_version: "34bfd03104a8a2ebd2e03f60c90353581cc5016a"
+          provenance_version: "v1.15.0-rc2"
```
Comment (Taztingo) [inline .github/workflows/test.yml:61]: Same as above.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -101,11 +101,11 @@ jobs:
           cd ./contracts/msgfees
           make all
       - name: Provenance Test
-        uses: provenance-io/provenance-testing-action@34bfd03104a8a2ebd2e03f60c90353581cc5016a
+        uses: provenance-io/provenance-testing-action@eeeaf2792a9fbab3ccd74439a9e132d079fcdef5
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
-          provenance_version: "v1.14.1"
-          smart_contract_action_version: "34bfd03104a8a2ebd2e03f60c90353581cc5016a"
+          provenance_version: "v1.15.0-rc2"
```
Comment (Taztingo) [inline .github/workflows/test.yml:107]: Same as above.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -124,11 +124,11 @@ jobs:
           cd ./contracts/name
           make all
       - name: Smart Contract Test setup
-        uses: provenance-io/provenance-testing-action@34bfd03104a8a2ebd2e03f60c90353581cc5016a
+        uses: provenance-io/provenance-testing-action@eeeaf2792a9fbab3ccd74439a9e132d079fcdef5
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
-          provenance_version: "v1.14.1"
-          smart_contract_action_version: "34bfd03104a8a2ebd2e03f60c90353581cc5016a"
+          provenance_version: "v1.15.0-rc2"
```
Comment (Taztingo) [inline .github/workflows/test.yml:130]: Same as above.

File: contracts/attrs/src/contract.rs [hunk 5]
```diff
@@ -164,423 +176,415 @@ fn try_update_label(
...
-        assert_eq!(rep.labels.len(), 1);
-        assert_eq!(
-            rep.labels[0],
-            Label {
-                text: "text".into(),
-            }
-        )
-    }
-}
+//
```
Comment (Taztingo) [inline contracts/attrs/src/contract.rs:198]: All the tests are commented out. Is this a mistake, or are we planning to introduce them at a different stage?

File: contracts/marker/src/contract.rs [hunk 6]
```diff
@@ -206,410 +240,400 @@ fn try_transfer(
...
-            denom: "nugz".into(),
-        };
-        let bin = query(deps.as_ref(), mock_env(), req).unwrap();
-        let marker: Marker = from_binary(&bin).unwrap();
-        assert_eq!(marker, expected_marker);
-        assert_eq!(marker.address, "tp18vmzryrvwaeykmdtu6cfrz5sau3dhc5c73ms0u")
-    }
-}
+//
+// #[cfg(test)]
```
Comment (Taztingo) [inline contracts/marker/src/contract.rs:265]: Same as above.

File: packages/provwasm-std-derive/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# osmosis-std-derive
```
Comment (Taztingo) [inline packages/provwasm-std-derive/README.md:1]: Very minor, but do we need to rename this stuff?

File: Makefile [hunk 8]
```diff
@@ -1,22 +1,17 @@
...
 .PHONY: contracts
 contracts:
 	@make -C contracts
 
 .PHONY: clean
 clean:
-	@make -C packages/bindings clean
-	@make -C packages/mocks clean
+	@make -C packages/provwasm-std clean
 	@make -C contracts clean
```
Comment (Taztingo) [inline Makefile:15]: Is it possible to have a make target that calls update-and-rebuild.sh for us? Might make it easier for people to know about that script. Would we want to tie it into `Make all` or would that slow it down too much?

**Review by Taztingo — CHANGES_REQUESTED** [review]

**Comment by Taztingo** [comment]
This looks great!

File: contracts/attrs/src/contract.rs (hunk 5 above)
Comment (kwtalley) [reply contracts/attrs/src/contract.rs:215]: most are back now, but will probably catch the last contract or 2 in the next RC

File: contracts/marker/src/contract.rs (hunk 6 above)
Comment (kwtalley) [reply contracts/marker/src/contract.rs:265]: most are back now, but will probably catch the last contract or 2 in the next RC

File: packages/provwasm-std-derive/README.md (hunk 7 above)
Comment (kwtalley) [reply packages/provwasm-std-derive/README.md]: yep, it's been removed in favor of a rewrite

File: Makefile (hunk 8 above)
Comment (kwtalley) [reply Makefile:15]: Yeah, that's the next work to do, right now there were a couple manual changes to the generated types so I'll need to get that moved into the generator code and then we can automate everything. but won't be part of this PR

**Pull request merged**